traits/has_traits.py

```diff
@@ -396,6 +419,16 @@ def update_traits_class_dict(class_name, bases, class_dict):
     # Move all trait definitions from the class dictionary to the
     # appropriate trait class dictionaries:
     for name, value in list(class_dict.items()):
+
+        # Warn name collisions.
+        if check_name_collision:
+            prefix = _disallowed_prefix(name)
+            if prefix:
+                warnings.warn("The attribute named '{}' of class {} uses a "
+                              "prefix '{}' that is discouraged. Consider "
+                              "renaming it." .format(name, class_name, prefix),
+                              UserWarning, 3)
```

Comment:
I am sorry for the misunderstanding: I did not mean to show *the matched prefix*. I was hoping to see all the protected prefixes, namely, `"trait"` and `"_trait"`.  This would save you having to loop over the disallowed prefixes as well.

What you had before was closer...something like:
```
if check_name_collision and name.startswith(DisallowedNamePrefixes):
    warnings.warn("The attribute named '{}' of class {} uses one of these prefixes: {}. Consider renaming it.".format(
name, class_name, ", ".join(DisallowedNamePrefixes))
```

Just mentioning here for future consideration: Consider checking `if prefix is None` instead of just `if prefix`, the former is more explicit and would not be confused with an empty string (which would likely be a programming error).